File: packages/core/src/platform/services/activityService.ts
import type { ActivityEvent, Page, PageQuery } from "@agently/contracts";
import type { PlatformDeps } from "./context.js";
import { paginate, byNewest } from "../util.js";

export interface ActivityService {
  list(query: PageQuery): Page<ActivityEvent>;
}

export function createActivityService(deps: PlatformDeps): ActivityService {
  const { repos } = deps;
  return {
    list(query) {
      const items = repos.activity.all().sort(byNewest((a) => a.at));
      return paginate(items, query);
    },
  };
}

File: apps/api/internal/services/activity_service.go
package services

import (
	"github.com/agently/api/internal/domain"
	"github.com/agently/api/internal/platform"
)

// ActivityService returns the recent workspace activity feed.
type ActivityService struct {
	deps Deps
}

func NewActivityService(deps Deps) *ActivityService {
	return &ActivityService{deps: deps}
}

func (s *ActivityService) List(query domain.PageQuery) domain.Page[domain.ActivityEvent] {
	items := s.deps.Repos.Activity.All()
	platform.SortByNewest(items, func(a domain.ActivityEvent) string { return string(a.At) })
	return platform.Paginate(items, query)
}
